### ISSUE TITLE
Proposal: Make Stream.CopyTo virtual

### ISSUE THREAD

**Issue by jamesqo** [dotnet/corefx#11342]
## Background

Currently for non-seekable streams, the overload of [`Stream.CopyTo`](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/Stream.cs#L191) that does not accept a buffer size defaults to using a 81920-length byte buffer between reads/writes. Although this is under the LOH threshold which is ~85K, it still generates substantial garbage for the GC to clean up, especially for people who are using small streams and aren't expecting such a gigantic buffer to be used.
## Proposal

We should make `Stream.CopyTo(Stream, int)` virtual. This would allow `byte[]`-based Streams, such as `MemoryStream`, to avoid using an intermediary buffer altogether, resulting in substantial memory savings.

The corresponding async version, `CopyToAsync(Stream, int, CancellationToken)`, is already virtual, so IMO there's no reason it shouldn't be done for this overload as well.
## Proposed API

``` cs
namespace System.IO
{
    public abstract class Stream
    {
        public void CopyTo(Stream destination, int bufferSize); // existing
        public virtual void CopyTo(Stream destination, int bufferSize); // new
    }

    public class MemoryStream : Stream
    {
        public override void CopyTo(Stream destination, int bufferSize);
    }
}
```

**Comment by justinvp** [comment]
API design regret: I wish I had specified this as virtual when we were adding it to .NET 4 to begin with. Luckily, in this case, changing it to virtual is a compatibile change, so this is one regret I won't have to harbor forever. Better late than never!

**Comment by benaadams** [comment]
Would overriding this method in a derived stream cause issue if used against an implementation that didn't have it as virtual? (e.g. or would the override be ignored and all be well?)

i.e. would it need harmonisation (and highlighting) at .Net Standard 1.X and implementation on derived stream behind a version `#ifdef`?

**Comment by jamesqo** [comment]
@benaadams Hmm, I just tested using csc. **TL;DR** It just ignores the override if it's no longer virtual, there's no errors thrown.

```
> cat Foo.cs
using System;

namespace Foo
{
    public class Program
    {
        public virtual void Blah()
        {
            Console.WriteLine("Base Blah");
        }
    }
}

> cat Bar.cs
using System;

namespace Bar
{
    public class E
    {
        public static void Main()
        {
            new P().Blah();
        }
    }

    public class P : Foo.Program
    {
        public override void Blah()
        {
            Console.WriteLine("Derived Blah");
        }
    }
}

> csc /out:Foo.dll /target:library Foo.cs

> csc Bar.cs /target:exe /reference:Foo.dll

> ./Bar.exe
Derived Blah

> sed -i 's/virtual //g' Foo.cs # I actually used vim here

> csc /out:Foo.dll /target:library Foo.cs

> ./Bar.exe
Base Blah

```

**Comment by benaadams** [comment]
Sounds ideal :+1:

**Comment by terrajobst** [comment]
This looks good as proposed.  Are there other stream types that should override `CopyTo`?

**Comment by jamesqo** [comment]
@terrajobst `BufferStream` would be a good candidate so we can avoid double buffering. At a quick glance through this repo `DeflateStream` and `FileStream` also look like they could use this, since they have their own private buffers.

**Comment by ianhays** [comment]
resolved by @jamesqo in https://github.com/dotnet/coreclr/pull/7070

**Comment by jamesqo** [comment]
@ianhays There are issues for some APIs which have been implemented but are still open, since they aren't yet exposed-- see [here](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20label%3Aapi-needs-exposed).

You may want to reopen the issue and label it with `api-needs-exposed`.

**Comment by justinvp** [comment]
Do we need an issue tracking making it virtual in the contract?

**Comment by jamesqo** [comment]
@justinvp IMO this issue should simply be reopened and tagged with the appropriate label, instead of creating a new one. Besides, I haven't yet submitted PRs for overriding the method in `BufferedStream` and `MemoryStream`.

**Comment by ianhays** [comment]
I was going to open a new one, but sure, we can use this one instead.

**Comment by karelz** [comment]
@jamesqo do you plan to expose the APIs?

**Comment by jamesqo** [comment]
@karelz I'm already doing so actually, see #12598

**Comment by karelz** [comment]
Great, assigning to you then ...

**Comment by jkotas** [comment]
Fixed by https://github.com/dotnet/corefx/pull/12598

**Comment by jamesqo** [comment]
@jkotas There are still some other places that need overriding, e.g. `BufferedStream` which is waiting on #13379 to be merged first. Have been super slow in sending PRs for this, sorry about that.

**Comment by karelz** [comment]
It seems to me that modifying `BufferedStream` is additional perf optimization, not part of exposure of the API. I would suggest to close this issue as finished and file a new one for tracking BufferedStream improvement.

**Comment by jamesqo** [comment]
@karelz, sure. In fact I believe this is already tracked by https://github.com/dotnet/corefx/issues/12860.
